js/app/widgets/dynamicLogo.js

```diff
@@ -0,0 +1,301 @@
+        if (['none', 'uppercase', 'lowercase'].indexOf(this._text_transform) == -1) {
+            let error = new Error('Unrecognized style property value for EknDynamicLogo-text-transform ' + this._text_transform);
+            logError(error);
+        }
+        this._update_text();
+     },
+
+    _update_text: function () {
+        let transformed;
+        switch (this._text_transform) {
+            case 'uppercase':
+                transformed = this._text.toLocaleUpperCase();
+                break;
+            case 'lowercase':
+                transformed = this._text.toLocaleLowerCase();
+                break;
+            default:
+                transformed = this._text;
+        }
+        this._layout.set_text(transformed, -1);
+    },
+
+    _get_fresh_state_context: function () {
+        let context = this.get_style_context();
+        let flags = this.get_state_flags();
+        context.save();
+        context.set_state(flags);
+        return context;
+    },
+
+    _get_margin: function () {
+        let context = this._get_fresh_state_context();
+        let margin = context.get_margin(context.get_state());
+        context.restore();
+        return margin;
+    },
+
+    vfunc_get_preferred_width: function () {
+        let min_width = EosKnowledgePrivate.style_context_get_int(this.get_style_context(),
+                                                                  'min-width',
+                                                                  this.get_state_flags());
+        min_width = min_width ? min_width : this.width_request;
+        min_width = min_width > 0 ? min_width : Width.MIN;
+        let nat_width = this._sizing === 'size-min' ? min_width : Width.MAX;
+
+        let margin = this._get_margin();
+        let extra = margin.left+ margin.right;
+
+        return [min_width + extra, nat_width + extra];
+    },
+
+    vfunc_get_preferred_height: function () {
+        let min_height = EosKnowledgePrivate.style_context_get_int(this.get_style_context(),
+                                                                   'min-height',
+                                                                   this.get_state_flags());
+        min_height = min_height ? min_height : this.height_request;
+        min_height = min_height > 0 ? min_height : Height.MIN;
+        let nat_height = this._sizing === 'size-min' ? min_height : Height.MAX;
+
+        let margin = this._get_margin();
+        let extra = margin.top + margin.bottom;
+
+        return [min_height + extra, nat_height + extra];
+    },
+
+    vfunc_size_allocate: function (alloc) {
+        let margin = this._get_margin();
+        alloc.x += margin.left;
+        alloc.y += margin.top;
+        alloc.width -= margin.left + margin.right;
+        alloc.height -= margin.top + margin.bottom;
+        this.set_allocation(alloc);
+    },
+
+    vfunc_draw: function (cr) {
+        let alloc = this.get_allocation();
```

Comment:
… But you should remove the margin from the allocation inside the `draw()` virtual function.

The reason why you should do it this way is to allow caching paths inside GTK+ to take into account the fact that the margin exists, and avoid a relayout if none is needed.

If the CSS margins change without changing the size of the allocation, GTK+ will just queue a redraw.